src/DependencyInjection/Compiler/RegisterHooksPass.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * Copyright (c) 2005-2017 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\CoreBundle\DependencyInjection\Compiler;
+
+use Symfony\Component\Config\Definition\Exception\InvalidConfigurationException;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+/**
+ * Find hook services and store them in an parameter.
+ *
+ * @author David Molineus <https://github.com/dmolineus>
+ */
+class RegisterHooksPass implements CompilerPassInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container)
+    {
+        $serviceIds = $container->findTaggedServiceIds('contao.hook');
+        $hooks      = [];
+
+        foreach ($serviceIds as $serviceId => $tags) {
+            foreach ($tags as $attributes) {
+                $this->guardRequiredAttributesExist($serviceId, $attributes);
+
+                $priority = $attributes['priority'] ?? 0;
+                $hook     = $attributes['hook'];
+
+                $hooks[$hook][$priority][] = [$serviceId, $attributes['method']];
+            }
+        }
+
+        if (count($hooks) > 0) {
+            // Apply priority sorting.
+            krsort($hooks);
```

Comment:
are you sure this works? Because it sorts on the key of the hook name, not the priority? Also, you might want to use the `\SplPriorityQueue` which is exactly meant for this.

Reply:
Actually I was because I craeted unit tests for it. But I had a closer look. It didn't work. `assertEquals` does not detect order of keys which I thought it would (and which I was used to do by PHPSpec which I actually prefer).

Fixed it in 802350f. Just used simple arrays as Symfony does the same in the event dispatcher.